Feat: show only relevant columns (as DataFrames) when unit tests fail

georgesittas opened the pull request:
Addresses the second item in https://github.com/TobikoData/sqlmesh/issues/1637. Below I've shared some examples to demonstrate how this looks, both in the CLI and the UI.

TL;DR instead of `difflib.ndiff` the diff will now be produced with `DataFrame.compare`.

EDIT: check out https://github.com/TobikoData/sqlmesh/pull/1741#issuecomment-1821065762 for updated views of the error reporting.

<details>
<summary>A single column in a single row is incorrect (outdated)</summary>

CLI:
```
➜  tests sqlmesh test
F
======================================================================
FAIL: test_example_full_model (tests/test_full_model.yaml)
----------------------------------------------------------------------
AssertionError: Data differs

  num_orders
    expected actual
0        5.0    2.0

----------------------------------------------------------------------
Ran 1 test in 0.013s

FAILED (failures=1)
```

UI:

<img width="544" alt="Screenshot 2023-11-20 at 8 26 14 PM" src="https://github.com/TobikoData/sqlmesh/assets/46752250/ef8770a0-daaa-48ba-b7c4-bdd99d8712bf">
</details>

<details>
<summary>Two columns in a single row are incorrect (outdated)</summary>

CLI:
```
➜  tests sqlmesh test
F
======================================================================
FAIL: test_example_full_model (tests/test_full_model.yaml)
----------------------------------------------------------------------
AssertionError: Data differs

   item_id        num_orders
  expected actual   expected actual
0      5.0    1.0        5.0    2.0

----------------------------------------------------------------------
Ran 1 test in 0.013s

FAILED (failures=1)
```

UI:

<img width="538" alt="Screenshot 2023-11-20 at 8 35 05 PM" src="https://github.com/TobikoData/sqlmesh/assets/46752250/055c35b9-0efa-42a5-b618-0e6245f07f2f">
</details>

<details>
<summary>Two columns in two different rows are incorrect (outdated)</summary>

CLI:
```
➜  tests sqlmesh test
F
======================================================================
FAIL: test_example_full_model (tests/test_full_model.yaml)
----------------------------------------------------------------------
AssertionError: Data differs

   item_id        num_orders
  expected actual   expected actual
0      NaN    NaN        5.0    2.0
1      5.0    2.0        NaN    NaN

----------------------------------------------------------------------
Ran 1 test in 0.013s

FAILED (failures=1)
```

UI:

<img width="503" alt="Screenshot 2023-11-20 at 9 27 44 PM" src="https://github.com/TobikoData/sqlmesh/assets/46752250/76816e45-bf33-42c4-9762-4bee30504d74">
</details>

<details>
<summary>Five columns in a single row are incorrect (outdated)</summary>

CLI:
```
➜  sushi git:(jo/improve_test_failure_msg) ✗ sqlmesh test
F.
======================================================================
FAIL: test_order_items (sushi/tests/test_order_items.yaml)
----------------------------------------------------------------------
AssertionError: Data differs

        id        order_id         item_id        quantity                 ds
  expected actual expected actual expected actual expected actual    expected      actual
0     10.0    1.0     10.0    1.0     10.0    2.0     10.0    4.0  2022-01-06  2022-01-01

----------------------------------------------------------------------
Ran 2 tests in 0.048s

FAILED (failures=1)
```

UI:

<img width="875" alt="Screenshot 2023-11-20 at 9 31 42 PM" src="https://github.com/TobikoData/sqlmesh/assets/46752250/c6dae6e5-acb8-42d0-a9b6-621e93df5e71">
</details>

References:
- https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.compare.html
- https://pandas.pydata.org/docs/user_guide/advanced.html#renaming-names-of-an-index-or-multiindex

z3z1ma commented on sqlmesh/core/test/definition.py:
```diff
@@ -111,13 +110,8 @@ def assert_equal(self, expected: pd.DataFrame, actual: pd.DataFrame) -> None:
-                )
-            )
-            e.args = (f"Data differs\n{diff}",)
+            diff = expected.compare(actual).rename(columns={"self": "expected", "other": "actual"})
```
Would it make sense to transpose beyond a certain number of columns?

So diff.T

georgesittas replied:
Maybe we should just transpose by default. Are you concerned about producing cluttered output for wide tables due to wrapping up the lines?

georgesittas replied:
For example:
```
F.
======================================================================
FAIL: test_order_items (sushi/tests/test_order_items.yaml)
----------------------------------------------------------------------
AssertionError: Data differs

                            0
id       expected        10.0
         actual           1.0
order_id expected        10.0
         actual           1.0
item_id  expected        10.0
         actual           2.0
quantity expected        10.0
         actual           4.0
ds       expected  2022-01-06
         actual    2022-01-01

----------------------------------------------------------------------
Ran 2 tests in 0.063s

FAILED (failures=1)
```

georgesittas replied:
OTOH if there are many rows that differ than we'll get the same line wrapping

georgesittas replied:
FWIW I'm not sure if the previous approach (i.e. using `difflib`) addressed these issues gracefully either. It would just spit out the whole rows which could line wrap pretty badly and also include unnecessary columns for which the actual and expected values matched.

georgesittas replied:
Nvm I forgot that Pandas doesn't wrap the lines but instead just hides the columns:

```
➜  test sqlmesh test
F
======================================================================
FAIL: test_example_full_model (tests/test_full_model.yaml)
----------------------------------------------------------------------
AssertionError: Data differs

   item_id        num_orders               a               b  ...      i        j               k               l
  expected actual   expected actual expected actual expected  ... actual expected actual expected actual expected actual
0      NaN    NaN        3.0    2.0     None      3     None  ...     11     None     12     None     13     None     14
1      1.0    2.0        NaN    NaN     None      3     None  ...     11     None     12     None     13     None     14

[2 rows x 28 columns]

----------------------------------------------------------------------
Ran 1 test in 0.024s

FAILED (failures=1)
```

@tobymao any thoughts on this?

tobymao commented:
it’s fine to truncate

tobymao commented:
it can be controlled with a flag

tobymao commented:
allow theh flag to be passed in

georgesittas commented on sqlmesh/core/test/definition.py:
```diff
@@ -95,10 +94,10 @@ def assert_equal(self, expected: pd.DataFrame, actual: pd.DataFrame) -> None:
+        # but if the actual is US, it doesn't take effect until the 2nd try!
         actual_types = actual.dtypes.to_dict()
-        expected = expected.astype(actual_types).astype(actual_types)
+        expected = expected.astype(actual_types, errors="ignore").astype(actual_types, errors="ignore")
```
FYI: the `errors="ignore"` were added here (mostly) because of the following case: an expected output column is `None` when the actual column is non-`None`. I was getting this error (which is also true in main):

```
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'NoneType'
```

Reference for the `errors` kwarg: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.astype.html

georgesittas replied:
It seemed in general we might not wanna fail in the `astype` and report the related DataFrame error, but instead let it go to the df comparison stage and show the diff itself? Let me know if this assumption doesn't look right.

georgesittas commented:
> allow the flag to be passed in

```
➜  test sqlmesh test
F
======================================================================
FAIL: test_example_full_model (tests/test_full_model.yaml)
----------------------------------------------------------------------
AssertionError: Data differs

   item_id        num_orders               a               b               c               d  ...      g        h               i               j               k               l
  expected actual   expected actual expected actual expected actual expected actual expected  ... actual expected actual expected actual expected actual expected actual expected actual
0      NaN    NaN        3.0    2.0     None      3     None      4     None      5     None  ...      9     None     10     None     11     None     12     None     13     None     14
1      1.0    2.0        NaN    NaN     None      3     None      4     None      5     None  ...      9     None     10     None     11     None     12     None     13     None     14

[2 rows x 28 columns]

----------------------------------------------------------------------
Ran 1 test in 0.024s

FAILED (failures=1)
➜  test sqlmesh test --no-truncate
F
======================================================================
FAIL: test_example_full_model (tests/test_full_model.yaml)
----------------------------------------------------------------------
AssertionError: Data differs

   item_id        num_orders               a               b               c  \
  expected actual   expected actual expected actual expected actual expected
0      NaN    NaN        3.0    2.0     None      3     None      4     None
1      1.0    2.0        NaN    NaN     None      3     None      4     None

                d               e               f               g         \
  actual expected actual expected actual expected actual expected actual
0      5     None      6     None      7     None      8     None      9
1      5     None      6     None      7     None      8     None      9

         h               i               j               k               l  \
  expected actual expected actual expected actual expected actual expected
0     None     10     None     11     None     12     None     13     None
1     None     10     None     11     None     12     None     13     None


  actual
0     14
1     14

----------------------------------------------------------------------
Ran 1 test in 0.018s

FAILED (failures=1)
```

georgesittas commented:
If the changes look good I can update the docs as well.

tobymao commented:
maybe just do -vv or verbose, that's what pytest does right?

additionally, there's too much expected vs actual, maybe we could abberivate it? exp. act. or something

tobymao commented on sqlmesh/cli/main.py:
```diff
@@ -364,10 +364,27 @@ def dag(ctx: click.Context, file: str) -> None:
+    no_truncate: bool,
+) -> None:
     """Run model unit tests."""
+    if no_truncate:
```
this needs to be done in the test so it can be reused, you need to add this interface to context, and notebooks as well right?

georgesittas replied:
Yeah good point, I'll take a look and see if I can reuse the `-v` flag otherwise I'll address this as well.

georgesittas commented:
> maybe just do -vv or verbose, that's what pytest does right?

I'm blind -- we already have a `-v` (for verbose) flag, not sure what it's used for. Maybe we can reuse that instead.

georgesittas commented:
> additionally, there's too much expected vs actual, maybe we could abberivate it? exp. act. or something, or maybe it's fine

I'm fine either way, can try your suggestion out and see how it looks. I defaulted to `expected` and `actual` because they seemed easier to parse

georgesittas commented:
Updated to use `-v` instead of `--no-truncate` since we already had that option. The `max_columns` pandas option is now set in `Context.test` as suggested. Also switched to use `exp` and `act` for `expected` and `actual`, respectively, but left a small note in the test failure message that lets the user know what they stand for.

Currently the test reporting looks like this:

```
➜  test sqlmesh test
F
======================================================================
FAIL: test_example_full_model (test/tests/test_full_model.yaml)
----------------------------------------------------------------------
AssertionError: Data differs (exp: expected, act: actual)

  item_id      num_orders          a         b         c         d         e         f         g         h         i         j         k         l
      exp  act        exp  act   exp act   exp act   exp act   exp act   exp act   exp act   exp act   exp act   exp act   exp act   exp act   exp act
0     NaN  NaN        3.0  2.0  None   3  None   4  None   5  None   6  None   7  None   8  None   9  None  10  None  11  None  12  None  13  None  14
1     1.0  2.0        NaN  NaN  None   3  None   4  None   5  None   6  None   7  None   8  None   9  None  10  None  11  None  12  None  13  None  14

----------------------------------------------------------------------
Ran 1 test in 0.023s

FAILED (failures=1)
```

And the verbose mode:

```
➜  test sqlmesh test -v
test_example_full_model (test/tests/test_full_model.yaml) ... FAIL

======================================================================
FAIL: test_example_full_model (test/tests/test_full_model.yaml)
----------------------------------------------------------------------
AssertionError: Data differs (exp: expected, act: actual)

  item_id      num_orders          a         b         c         d         e  \
      exp  act        exp  act   exp act   exp act   exp act   exp act   exp
0     NaN  NaN        3.0  2.0  None   3  None   4  None   5  None   6  None
1     1.0  2.0        NaN  NaN  None   3  None   4  None   5  None   6  None

          f         g         h         i         j         k         l
  act   exp act   exp act   exp act   exp act   exp act   exp act   exp act
0   7  None   8  None   9  None  10  None  11  None  12  None  13  None  14
1   7  None   8  None   9  None  10  None  11  None  12  None  13  None  14

----------------------------------------------------------------------
Ran 1 test in 0.018s

FAILED (failures=1)
```
